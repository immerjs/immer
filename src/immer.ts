import {
	applyPatchesImpl,
	createDraftImpl,
	createImmerContext,
	Draft,
	finishDraftImpl,
	Immer,
	Immutable,
	IProduce,
	IProduceWithPatches,
	Objectish,
	Patch,
	PatchListener,
	produceImpl,
	produceWithPatchesImpl,
	setAutoFreezeImpl,
	setUseStrictShallowCopyImpl
} from "./internal"

export {
	Draft,
	WritableDraft,
	Immutable,
	Patch,
	PatchListener,
	Producer,
	original,
	current,
	isDraft,
	isDraftable,
	NOTHING as nothing,
	DRAFTABLE as immerable,
	freeze,
	Objectish,
	StrictMode
} from "./internal"

const globalImmerContext = createImmerContext()

/**
 * The `produce` function takes a value and a "recipe function" (whose
 * return value often depends on the base state). The recipe function is
 * free to mutate its first argument however it wants. All mutations are
 * only ever applied to a __copy__ of the base state.
 *
 * Pass only a function to create a "curried producer" which relieves you
 * from passing the recipe function every time.
 *
 * Only plain objects and arrays are made mutable. All other objects are
 * considered uncopyable.
 *
 * Note: This function is __bound__ to its `Immer` instance.
 *
 * @param {any} base - the initial state
 * @param {Function} producer - function that receives a proxy of the base state as first argument and which can be freely modified
 * @param {Function} patchListener - optional function that will be called with all the patches produced here
 * @returns {any} a new state, or the initial state if nothing was modified
 */
export const produce: IProduce = produceImpl.bind(globalImmerContext)

/**
 * Like `produce`, but `produceWithPatches` always returns a tuple
 * [nextState, patches, inversePatches] (instead of just the next state)
 */
export const produceWithPatches: IProduceWithPatches = produceWithPatchesImpl.bind(
	globalImmerContext
)

/**
 * Pass true to automatically freeze all copies created by Immer.
 *
 * Always freeze by default, even in production mode
 */
export const setAutoFreeze = setAutoFreezeImpl.bind(globalImmerContext)

/**
 * Pass true to enable strict shallow copy.
 *
 * By default, immer does not copy the object descriptors such as getter, setter and non-enumrable properties.
 */
export const setUseStrictShallowCopy = setUseStrictShallowCopyImpl.bind(
	globalImmerContext
)

/**
 * Apply an array of Immer patches to the first argument.
 *
 * This function is a producer, which means copy-on-write is in effect.
 */
export const applyPatches = applyPatchesImpl.bind(globalImmerContext) as <
	T extends Objectish
>(
	base: T,
	patches: readonly Patch[]
) => T

/**
 * Create an Immer draft from the given base state, which may be a draft itself.
 * The draft can be modified until you finalize it with the `finishDraft` function.
 */
export const createDraft = createDraftImpl.bind(globalImmerContext) as <
	T extends Objectish
>(
	base: T
) => Draft<T>

/**
 * Finalize an Immer draft from a `createDraft` call, returning the base state
 * (if no changes were made) or a modified copy. The draft must *not* be
 * mutated afterwards.
 *
 * Pass a function as the 2nd argument to generate Immer patches based on the
 * changes that were made.
 */
export const finishDraft = finishDraftImpl.bind(globalImmerContext) as <
	D extends Draft<any>
>(
	draft: D,
	patchListener?: PatchListener
) => D extends Draft<infer T> ? T : never

/**
 * This function is actually a no-op, but can be used to cast an immutable type
 * to an draft type and make TypeScript happy
 *
 * @param value
 */
export function castDraft<T>(value: T): Draft<T> {
	return value as any
}

/**
 * This function is actually a no-op, but can be used to cast a mutable type
 * to an immutable type and make TypeScript happy
 * @param value
 */
export function castImmutable<T>(value: T): Immutable<T> {
	return value as any
}

export {Immer}

export {enablePatches} from "./plugins/patches"
export {enableMapSet} from "./plugins/mapset"
