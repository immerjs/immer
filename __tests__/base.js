import immer from ".."

describe("base", () => {
  let baseState
  let origBaseState

  beforeEach(() => {
    origBaseState = baseState = createBaseState()
  })

  it("should return the original without modifications", () => {
    const nextState = immer(baseState, () => {})
    expect(nextState).toBe(baseState)
  })

  it("should return the original without modifications when reading stuff", () => {
    const nextState = immer(baseState, s => {
      expect(s.aProp).toBe("hi")
      expect(s.anObject.nested).toEqual({yummie: true})
    })
    expect(nextState).toBe(baseState)
  })

  it("should not return any value: thunk", () => {
    const warning = jest.spyOn(console, "warn")
    immer(baseState, () => ({bad: "don't do this"}))
    immer(baseState, () => [1, 2, 3])
    immer(baseState, () => false)
    immer(baseState, () => "")

    expect(warning).toHaveBeenCalledTimes(4)
  })

  it("should return a copy when modifying stuff", () => {
    const nextState = immer(baseState, s => {
      s.aProp = "hello world"
    })
    expect(nextState).not.toBe(baseState)
    expect(baseState.aProp).toBe("hi")
    expect(nextState.aProp).toBe("hello world")
    // structural sharing?
    expect(nextState.nested).toBe(baseState.nested)
  })

  it("deep change bubbles up", () => {
    const nextState = immer(baseState, s => {
      s.anObject.nested.yummie = false
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anObject).not.toBe(baseState.anObject)
    expect(baseState.anObject.nested.yummie).toBe(true)
    expect(nextState.anObject.nested.yummie).toBe(false)
    expect(nextState.anArray).toBe(baseState.anArray)
  })

  it("can add props", () => {
    const nextState = immer(baseState, s => {
      s.anObject.cookie = {tasty: true}
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anObject).not.toBe(baseState.anObject)
    expect(nextState.anObject.nested).toBe(baseState.anObject.nested)
    expect(nextState.anObject.cookie).toEqual({tasty: true})
  })

  it("can delete props", () => {
    const nextState = immer(baseState, s => {
      delete s.anObject.nested
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anObject).not.toBe(baseState.anObject)
    expect(nextState.anObject.nested).toBe(undefined)
  })

  it("ignores single non-modification", () => {
    const nextState = immer(baseState, s => {
      s.aProp = "hi"
    })
    expect(nextState).toBe(baseState)
  })

  it("processes single modification", () => {
    const nextState = immer(baseState, s => {
      s.aProp = "hello"
      s.aProp = "hi"
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState).toEqual(baseState)
  })

  it("should support reading arrays", () => {
    const nextState = immer(baseState, s => {
      s.anArray.slice()
    })
    expect(nextState.anArray).toBe(baseState.anArray)
    expect(nextState).toBe(baseState)
  })

  it("should support changing arrays", () => {
    const nextState = immer(baseState, s => {
      s.anArray[3] = true
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anArray).not.toBe(baseState.anArray)
    expect(nextState.anArray[3]).toEqual(true)
  })

  it("should support changing arrays - 2", () => {
    const nextState = immer(baseState, s => {
      s.anArray.splice(1, 1, "a", "b")
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anArray).not.toBe(baseState.anArray)

    expect(nextState.anArray).toEqual([3, "a", "b", {c: 3}, 1])
  })

  it("should support sorting arrays", () => {
    const nextState = immer(baseState, s => {
      s.anArray[2].c = 4
      s.anArray.sort()
      s.anArray[3].c = 5
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anArray).not.toBe(baseState.anArray)
    expect(nextState.anArray).toEqual([1, 2, 3, {c: 5}])
  })

  it("should updating inside arrays", () => {
    const nextState = immer(baseState, s => {
      s.anArray[2].test = true
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anArray).not.toBe(baseState.anArray)
    expect(nextState.anArray).toEqual([3, 2, {c: 3, test: true}, 1])
  })

  it("reusing object should work", () => {
    const nextState = immer(baseState, s => {
      debugger
      const obj = s.anObject
      delete s.anObject
      s.messy = obj
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anArray).toBe(baseState.anArray)
    expect(nextState).toEqual({
      anArray: [3, 2, {c: 3}, 1],
      aProp: "hi",
      messy: {
        nested: {
          yummie: true,
        },
        coffee: false,
      },
    })
    expect(nextState.messy.nested).toBe(baseState.anObject.nested)
  })

  it("refs should be transparent", () => {
    const nextState = immer(baseState, s => {
      const obj = s.anObject
      s.aProp = "hello"
      delete s.anObject
      obj.coffee = true
      s.messy = obj
    })
    expect(nextState).not.toBe(baseState)
    expect(nextState.anArray).toBe(baseState.anArray)
    expect(nextState).toEqual({
      anArray: [3, 2, {c: 3}, 1],
      aProp: "hello",
      messy: {
        nested: {
          yummie: true,
        },
        coffee: true,
      },
    })
    expect(nextState.messy.nested).toBe(baseState.anObject.nested)
  })

  it("should allow setting to undefined a defined draft property", () => {
    const nextState = immer(baseState, s => {
      s.aProp = undefined
    })
    expect(nextState).not.toBe(baseState)
    expect(baseState.aProp).toBe("hi")
    expect(nextState.aProp).toBe(undefined)
  })

  afterEach(() => {
    expect(baseState).toBe(origBaseState)
    expect(baseState).toEqual(createBaseState())
  })

  function createBaseState() {
    return {
      anArray: [3, 2, {c: 3}, 1],
      aProp: "hi",
      anObject: {
        nested: {
          yummie: true,
        },
        coffee: false,
      },
    }
  }
})

describe("readme example", () => {
  it("works", () => {
    const baseState = [
      {
        todo: "Learn typescript",
        done: true,
      },
      {
        todo: "Try immer",
        done: false,
      },
    ]

    const nextState = immer(baseState, state => {
      state.push({todo: "Tweet about it"})
      state[1].done = true
    })

    // the new item is only added to the next state,
    // base state is unmodified
    expect(baseState.length).toBe(2)
    expect(nextState.length).toBe(3)

    // same for the changed 'done' prop
    expect(baseState[1].done).toBe(false)
    expect(nextState[1].done).toBe(true)

    // unchanged data is structurally shared
    expect(nextState[0]).toBe(baseState[0])
    // changed data not (d√ªh)
    expect(nextState[1]).not.toBe(baseState[1])
  })
})
