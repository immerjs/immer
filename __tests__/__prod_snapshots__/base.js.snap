// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=false > map drafts > revokes map proxies 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=false > map drafts > revokes map proxies 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=false > recipe functions > cannot return a modified child draft 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=false > recipe functions > cannot return an object that references itself 1`] = `[RangeError: Maximum call stack size exceeded]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=false > set drafts > revokes sets 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=false > set drafts > revokes sets 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=false > throws when Object.defineProperty() is used on drafts 1`] = `[Error: [Immer] minified error nr: 11. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=false > throws when Object.setPrototypeOf() is used on a draft 1`] = `[Error: [Immer] minified error nr: 12. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=false > throws when the draft is modified and another object is returned 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=true > map drafts > revokes map proxies 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=true > map drafts > revokes map proxies 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=true > recipe functions > cannot return a modified child draft 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=true > recipe functions > cannot return an object that references itself 1`] = `[RangeError: Maximum call stack size exceeded]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=true > set drafts > revokes sets 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=true > set drafts > revokes sets 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=true > throws when Object.defineProperty() is used on drafts 1`] = `[Error: [Immer] minified error nr: 11. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=true > throws when Object.setPrototypeOf() is used on a draft 1`] = `[Error: [Immer] minified error nr: 12. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=false:use-listener=true > throws when the draft is modified and another object is returned 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=false > map drafts > revokes map proxies 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=false > map drafts > revokes map proxies 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=false > recipe functions > cannot return a modified child draft 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=false > recipe functions > cannot return an object that references itself 1`] = `[RangeError: Maximum call stack size exceeded]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=false > set drafts > revokes sets 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=false > set drafts > revokes sets 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=false > throws when Object.defineProperty() is used on drafts 1`] = `[Error: [Immer] minified error nr: 11. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=false > throws when Object.setPrototypeOf() is used on a draft 1`] = `[Error: [Immer] minified error nr: 12. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=false > throws when the draft is modified and another object is returned 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=true > map drafts > revokes map proxies 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=true > map drafts > revokes map proxies 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=true > recipe functions > cannot return a modified child draft 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=true > recipe functions > cannot return an object that references itself 1`] = `[RangeError: Maximum call stack size exceeded]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=true > set drafts > revokes sets 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=true > set drafts > revokes sets 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=true > throws when Object.defineProperty() is used on drafts 1`] = `[Error: [Immer] minified error nr: 11. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=true > throws when Object.setPrototypeOf() is used on a draft 1`] = `[Error: [Immer] minified error nr: 12. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=false:shallow-copy=true:use-listener=true > throws when the draft is modified and another object is returned 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=false > map drafts > revokes map proxies 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=false > map drafts > revokes map proxies 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=false > recipe functions > cannot return a modified child draft 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=false > recipe functions > cannot return an object that references itself 1`] = `[RangeError: Maximum call stack size exceeded]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=false > set drafts > revokes sets 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=false > set drafts > revokes sets 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=false > throws when Object.defineProperty() is used on drafts 1`] = `[Error: [Immer] minified error nr: 11. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=false > throws when Object.setPrototypeOf() is used on a draft 1`] = `[Error: [Immer] minified error nr: 12. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=false > throws when the draft is modified and another object is returned 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=true > map drafts > revokes map proxies 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=true > map drafts > revokes map proxies 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=true > recipe functions > cannot return a modified child draft 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=true > recipe functions > cannot return an object that references itself 1`] = `[RangeError: Maximum call stack size exceeded]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=true > set drafts > revokes sets 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=true > set drafts > revokes sets 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=true > throws when Object.defineProperty() is used on drafts 1`] = `[Error: [Immer] minified error nr: 11. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=true > throws when Object.setPrototypeOf() is used on a draft 1`] = `[Error: [Immer] minified error nr: 12. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=false:use-listener=true > throws when the draft is modified and another object is returned 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=false > map drafts > revokes map proxies 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=false > map drafts > revokes map proxies 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=false > recipe functions > cannot return a modified child draft 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=false > recipe functions > cannot return an object that references itself 1`] = `[RangeError: Maximum call stack size exceeded]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=false > set drafts > revokes sets 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=false > set drafts > revokes sets 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=false > throws when Object.defineProperty() is used on drafts 1`] = `[Error: [Immer] minified error nr: 11. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=false > throws when Object.setPrototypeOf() is used on a draft 1`] = `[Error: [Immer] minified error nr: 12. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=false > throws when the draft is modified and another object is returned 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=true > map drafts > revokes map proxies 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=true > map drafts > revokes map proxies 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=true > recipe functions > cannot return a modified child draft 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=true > recipe functions > cannot return an object that references itself 1`] = `[RangeError: Maximum call stack size exceeded]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=true > set drafts > revokes sets 1`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=true > set drafts > revokes sets 2`] = `[Error: [Immer] minified error nr: 3. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=true > throws when Object.defineProperty() is used on drafts 1`] = `[Error: [Immer] minified error nr: 11. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=true > throws when Object.setPrototypeOf() is used on a draft 1`] = `[Error: [Immer] minified error nr: 12. Full error at: https://bit.ly/3cXEKWf]`;

exports[`base functionality - auto-freeze=true:shallow-copy=true:use-listener=true > throws when the draft is modified and another object is returned 1`] = `[Error: [Immer] minified error nr: 4. Full error at: https://bit.ly/3cXEKWf]`;

exports[`complex nesting map / set / object > modify deep object 1`] = `
{
  "map": Map {
    "set1" => Set {
      {
        "a": 2,
      },
      {
        "b": 2,
      },
    },
    "set2" => Set {
      {
        "c": 3,
      },
    },
  },
}
`;

exports[`complex nesting map / set / object > modify deep object 2`] = `
[
  {
    "op": "remove",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 1,
    },
  },
  {
    "op": "add",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 2,
    },
  },
]
`;

exports[`complex nesting map / set / object > modify deep object 3`] = `
{
  "map": Map {
    "set1" => Set {
      {
        "a": 2,
      },
      {
        "b": 2,
      },
    },
    "set2" => Set {
      {
        "c": 3,
      },
    },
  },
}
`;

exports[`complex nesting map / set / object > modify deep object 4`] = `
[
  {
    "op": "remove",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 1,
    },
  },
  {
    "op": "add",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 2,
    },
  },
]
`;

exports[`complex nesting map / set / object > modify deep object 5`] = `
{
  "map": Map {
    "set1" => Set {
      {
        "a": 2,
      },
      {
        "b": 2,
      },
    },
    "set2" => Set {
      {
        "c": 3,
      },
    },
  },
}
`;

exports[`complex nesting map / set / object > modify deep object 6`] = `
[
  {
    "op": "remove",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 1,
    },
  },
  {
    "op": "add",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 2,
    },
  },
]
`;

exports[`complex nesting map / set / object > modify deep object 7`] = `
{
  "map": Map {
    "set1" => Set {
      {
        "a": 2,
      },
      {
        "b": 2,
      },
    },
    "set2" => Set {
      {
        "c": 3,
      },
    },
  },
}
`;

exports[`complex nesting map / set / object > modify deep object 8`] = `
[
  {
    "op": "remove",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 1,
    },
  },
  {
    "op": "add",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 2,
    },
  },
]
`;

exports[`complex nesting map / set / object > modify deep object 9`] = `
{
  "map": Map {
    "set1" => Set {
      {
        "a": 2,
      },
      {
        "b": 2,
      },
    },
    "set2" => Set {
      {
        "c": 3,
      },
    },
  },
}
`;

exports[`complex nesting map / set / object > modify deep object 10`] = `
[
  {
    "op": "remove",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 1,
    },
  },
  {
    "op": "add",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 2,
    },
  },
]
`;

exports[`complex nesting map / set / object > modify deep object 11`] = `
{
  "map": Map {
    "set1" => Set {
      {
        "a": 2,
      },
      {
        "b": 2,
      },
    },
    "set2" => Set {
      {
        "c": 3,
      },
    },
  },
}
`;

exports[`complex nesting map / set / object > modify deep object 12`] = `
[
  {
    "op": "remove",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 1,
    },
  },
  {
    "op": "add",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 2,
    },
  },
]
`;

exports[`complex nesting map / set / object > modify deep object 13`] = `
{
  "map": Map {
    "set1" => Set {
      {
        "a": 2,
      },
      {
        "b": 2,
      },
    },
    "set2" => Set {
      {
        "c": 3,
      },
    },
  },
}
`;

exports[`complex nesting map / set / object > modify deep object 14`] = `
[
  {
    "op": "remove",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 1,
    },
  },
  {
    "op": "add",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 2,
    },
  },
]
`;

exports[`complex nesting map / set / object > modify deep object 15`] = `
{
  "map": Map {
    "set1" => Set {
      {
        "a": 2,
      },
      {
        "b": 2,
      },
    },
    "set2" => Set {
      {
        "c": 3,
      },
    },
  },
}
`;

exports[`complex nesting map / set / object > modify deep object 16`] = `
[
  {
    "op": "remove",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 1,
    },
  },
  {
    "op": "add",
    "path": [
      "map",
      "set1",
      0,
    ],
    "value": {
      "a": 2,
    },
  },
]
`;
